### ISSUE TITLE
As a content editor I want to embed data from data.gov.au in a visual format so that my audience can more easily understand findings

### ISSUE THREAD

**Issue by stooit** [govCMS/GovCMS7#146]
We wish to build a data.gov.au bridge as a custom module, to help provide a simple 'click button' approach to embedding datasets as tables, graphs and maps within govCMS.

The idea is that datasets stored on data.gov.au (or more generally, _any_ CKAN service) can be easily referenced, annotated with local metadata, and displayed as a graph, map, or table.

General:
- Allow authenticated access to any CKAN API
- Local, configurable caching of datasets
- Fieldable entities to allow for localised metadata and extension
- Indexed and searchable entities

Graphing:
- Using d3js, flexible visualisation library
- Configurable colours for visualisation display
- Types:
  -- Bar (horizontal / vertical variants)
  -- Stacked bar (horizontal / vertical variants)
  -- Line
  -- Stacked area
  -- Box plot (horizontal / vertical variants)
  -- Pie

Tabular data display:
- Sortable columns
- Responsive display
- Searchable/filterable dataset

_note_: mapping capabilities are on the roadmap but the functional spec has not yet been scoped.

Embed:
- WYSIWYG plugin to allow easy embed and display within any filtered content

The above functionality is what we're aiming for as a MVP, but the module would be built to be as flexible and extensible as possible. Ideally this would allow for further graph types and display enhancements down the line.

**Comment by WebProject2015** [comment]
This is a great idea, but it does not go far enough.  

Have you given consideration to making the bridge bidirectional?  It would be perfect if govCMS could  also push dataset updates to data.gov.au.   

We currently import CSVs to govCMS and store each row as a node to allow searchable dynamic tables. Indeed the vast majority of nodes on our site are derived from structured datasets. We are also planning to add many many more such structured data imports.  If these were all references to data on data.gov.au our site would be a whole lot smaller and more performant. 

  The biggest barriers to publishing data to data.gov.au are; the difficulty of data transformation prior to import, the specialist knowledge required to set up a dataset and the publishing overhead of updating data in multiple locations.   All of these problems can be addressed via govCMS integration.    

 A bi-directional bridge would strongly align with the government's Digital Transformation Agenda, tick all the boxes in the Digital Services Standard and would enable agencies to combine and remix datasets and visualisations through govCMS.  This would be a powerful step for digital governance, improving govCMS sites, improving data.gov.au, improving government's access to interagency data and improving the user experience of government data.

**Comment by stooit** [comment]
@WebProject2015 We're aiming for MVP implementation to meet an immediate client need, which assumes data is already available via data.gov.au and focusses on initial API connectivity and display. Definitely agree with the gains to be had by allowing import to data.gov.au via govCMS.

Ideally we can build a sane, pluggable base so the opportunity to extend and enhance in future will be made available to all. The 'push' functionality sounds like the perfect candidate for a sub-module here.

**Comment by gavintapp** [comment]
@WebProject2015 that would be good - but might also be trying to boil the ocean :) data.gov.au has data import tools at present, and it could be argued that greater use of data.gov.au might be encouraged by easier ways of making use of the available data.

govCMS may not be the right place to solve the issue you describe of data transformation before adding to data.gov.au. As I understand it, govCMS is not intended for transactional services (like updating a dataset) but aims to be good at presenting content/interfaces from other systems.

**Comment by fiasco** [comment]
Agree with @gavintapp and @srowlands that this issue is about the ability to read-only from data.gov.au. Doesn't mean another issue can't be opened though to support write access to data.gov.au from govCMS. So @WebProject2015 please feel free to log that if thats something you'd like to see in govCMS.

**Comment by fiasco** [comment]
This solution can be broken into three components:
1. Store references to data.gov.au (CKAN) datasets. But not the data itself.
2. Provide a turn-key capability to embed data from CKAN in tabular format on a web page.
3. Using javascript libraries, enrich the tabular display by converting it into a graph or map based on the data context.

The first two components are great candidates for the govCMS distribution. This will simply allow agencies to display CSV data from data.gov.au as tabular data on their websites.

The third component has much greater complexities to implement for a distribution than an agency. 

Agencies have complete control over their data and can ensure the data displayed matches a format to suit their graphing and mapping visualisation requirements. 

To implement this for a distribution however requires deeper abilities to allow dynamic mapping of dataset columns to specific uses in mapping and graphing visualisations. Because this capability is not intended on being delivered at this stage, IMO, its not a good fit to add to govCMS. 

However, agencies can still achieve this in their theme. As an alternative to a govCMS contribution for this component, I suggest you deliver a blog post as an outcome on how to enrich data.gov.au data on govcms. If there is enough community movement around that, then perhaps at that point, this integration could be built into govCMS.

I've done some investigation of [CKAN module](http://drupal.org/project/ckan) with @srowlands . We've concluded its a bad move for govCMS to use this module due to the low quality of code and commitment from the maintainers. Also there is a small amount of overlap between functional needs of this issue and what the module provides.
**Pros**
- Can query for datasets (overlap with DoE requirement)
- Can push datasets to data.gov.au
- Using community software

**Cons**
- Poorly written
- Minimally maintained
- Needs work to be used in govCMS

Finally, data.gov.au is a public/unclass data service. API keys are required to access unpublished data. This may be desired if the dataset on data.gov.au and the govcms webpage displaying the data wish to be published at the sametime, then the API key will be required to retrieve data while the page is also in draft state.

So final decisions for @teamglenny / @aleayr :
- OK to exclude graphing and mapping of data.gov.au data from govCMS inplace of a blog post about how to do it?
- OK to not use CKAN module?
- OK to use API keys to access unpublished data?

**Comment by teamglenny** [comment]
In essence there is agreement and approval for the functionality to be developed.

We’ve had discussions with our TAM and technical experts and agree that a module is the way to go, in particular for the retrieval and presentation of tabular data.

In regards to graphical representation it is felt a base line of tools should be baked in, but not so much as to include everything and the kitchen sink as this may represent distribution and maintenance bloat. There is leeway in this comment though as the client is effectively sponsoring this as the “first” agency so the generic requirements can be built from the client requirements.  We rely on the developers to identify where the line in the sand is for this.

For anything that would be construed as “advanced graphical representation only ever relevant to DoE” there should be consideration for baking this into the client theme layer.

There has been community feedback it should also provide the ability to feed data back into data.gov.au.  Whilst this is something that could occur down the track its felt without an immediate strong use case it is beyond the scope of the client project and shouldn’t be considered for the MVP.

@srowlands please have a chat with @jozhao at earliest convenience.  He has done some work on a similar requirement for another agency which may be a starting point alternative for you

**tl:dr** new module ok, tabular data ok, graphical representation ok at a base level, any advanced graphical representation purely niche to client bake in at the theme layer or other alternative, feedback to data.gov.au not in this first round,

**Comment by tobybellwood** [comment]
From a data.gov.au perspective, the proposals to build modules to ingest/consume data.gov.au datasets make really good sense.  Having a reusable codebase for the extraction of machine readable datasets into GovCMS would be a boon for both platforms, plus getting these high-quality datasets into data.gov.au would be fantastic :-)

I’m not entirely sold on using GovCMS to upload to data.gov.au – I’d certainly prefer to see us provide better tools and more assistance to help data publishers with their use of the data.gov.au platform. There may be alternative solutions that we can work through if necessary?

Always happy for people to come talk to us, the data@pmc.gov.au email is the best starting point.

**Comment by teamglenny** [comment]
Thanks for the input, @tobybellwood :)

**Comment by Starl3n** [comment]
In CKAN datasets package together resources under a common description, inc metadata. If you wanted a module for table views then you'd want to target a specific resource or use a rule to select which resource(s) to embed.
CKAN already has embed ready resource views. As such, a wysiwyg addition only needs to let a user browse the list of datasets>resources for data.gov.au and then paste the relevant embed code.
Resource views already exist for Tabular, spatial and document resources on DAta.gov.au.

**Comment by Starl3n** [comment]
Btw, take a look at data.gov.sg for a CKAN portal that is bringing data vis up into the dataset page, group page and home page. There should be a few examples of this approach being used in Australia soonish.

**Comment by Wongad** [comment]
Looking forward to see if there's a solution for this be it the CKAN module or the Embed modules, as we've also some data from http://data.gov.au/organization/australian-financial-security-authority we would like to reuse on our site.

**Comment by stooit** [comment]
We will have an update on this issue shortly (along with a demo environment of sorts). Will keep you all posted as things progress.

**Comment by Starl3n** [comment]
What data are you trying to embed? The embed code can already be found on data.gov.au against each resource view...

Such as 

<iframe width="700" height="400" src="http://data.gov.au/dataset/business-non-business-personal-insolvency-statistics/resource/932db4f8-be84-43a8-ad5a-86c17f62a185/view/b5b92eed-fff9-4a7c-876e-e131fe91b407" frameBorder="0"></iframe>

**Comment by Starl3n** [comment]
code was removed from the comment :) but, easy to find against any resource...

**Comment by stooit** [comment]
@Starl3n it's ok, I edited your comment to peek at the embed code. ;)

We're aware we can embed via iFrame; what we're building is a way to more tightly integrate with govCMS to provide a 'wizard' style approach for enhanced ease of use, and a pluggable visualisation framework for more control over the resulting display.

The result should be more flexibility for developers who wish to build custom visualisations, and greater ease of use for people wishing to embed data in various formats within govCMS.

**Comment by Starl3n** [comment]
yep - just saying it would be better to do that work upstream in CKAN itself rather than Drupal :) Then any CMS can embed the visualizations...

**Comment by Wongad** [comment]
Thanks @Starl3n , it's not the issue with the embed code but more of how we could embed in govCMS or query/push dataset in data.gov.au which will lighten our work to update these stats.

The AFSA site is not in govCMS yet but we're in the midst of sorting out our priorities.

**Comment by murraywoodman** [comment]
I'll throw my 2c in here. When we were doing our work on the Govt Demo site http://government.site-showcase.com/we implemented the ability to display charts. 

We did this as follows:
- Paragraph bundle called "Chart"
- Textarea field to hold a pasted CSV
- Link field to hold a URL pointing to a CSV
- Select field to select a chart template
- A few other fields to hold common data for labels, units, headings etc

In the theme we implemented a template for the chart bundle which:
- got the data
- added Highcharts library
- added the chart template assets
- attached the data to the template

We managed to do all of this using govCMS out of the box. ie. Config, Content and Theme. No Modules.

By crafting some general purpose chart templates covering the usual suspects (Bar, Pie, etc) we got good results.

If you wanted to go crazy (fancy JS, loading streaming data etc), extra chart templates could be implemented in the theme layer by someone who knew JS.

You can see a demo page here:
http://government.site-showcase.com/demos/highcharts

WARNING: This page has rotted a bit in the time since we implemented it. The charts are not working properly. I am only showing this an an example of what is possible to give you an idea. What you are seeing on this page is a number of paragraph items using the approach above.

When we implemented this we wanted to give editors the ability to add data and get charts. I imagine that the approach could be used for CKAN / DKAN if you can get the data as a CSV.

What are the downsides?

Highcharts has a lot of config options. It is not practical to capture all of this config in a single entity (paragraph item). The templates must therefore make a fair number of assumptions in order to produce a chart. It is possible to hit a sweet spot for general use cases, however, specific requirements will require a new template to be produced in th ebackend. I therefore think that the approach described will work well for simple data. Large datasets, with unusual shapes or requirements will pose more of a challenge.

I'm happy to share the code we have worked on if this is of interest. As you can see it requires a few updates but it would form a good basis if you were interested in this kind of functionality.

**Comment by Wongad** [comment]
Thanks @murraywoodman , will definitely consider using what you've recommended if the module doesn't pass through.

**Comment by stooit** [comment]
Thanks for sharing @murraywoodman, always good to hear how others are tackling similar issues!

A few things: 
We wanted to build a 'looking glass' into data.gov.au as the canonical source, so we don't have to deal with duplication of data.

CKAN also provides a DataStore extension, which gives a powerful endpoint to query data. You could upload an extremely large CSV which gets translated into data that can be directly queried. As you mentioned you probably couldn't practically deal with large datasets (e.g a giant CSV into a long-text field), nor could you easily query for subsets of data within it (without client-side parsing).

We also wanted to promote reusability within govCMS; build it once and make it a low enough barrier to entry that everyone can benefit from it. We're building out a framework that (hopefully!) covers a large portion of basic use-cases, while others have the option to build on top of with their own custom visualisations. That way we aren't mandating the use of highcharts vs d3, it's abstracted so others can build on top of it in the way that suits them.

Thanks for the input, will post back with an update on progress in the near future.

**Comment by aleayr** [comment]
This is now in the distro as govcms_ckan module.
